### PR TITLE
Add Spring retry to S3S3copier

### PR DESCRIPTION
**Pull request by max-jacobs**
Fix for [issue 56](https://github.com/HotelsDotCom/circus-train/issues/56).

`RetrayableTransferManager` wraps `TransferManager.copy()` in order to provide a retry mechanism, implemented using Spring retry.

### CODE REVIEW

File: circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  public RetryableTransferManager(TransferManager transferManager, AmazonS3 srcClient) {
+    this.transferManager = transferManager;
+    this.srcClient = srcClient;
+  }
+
+  @Retryable(value = { AmazonS3Exception.class },
+      backoff = @Backoff(delay = BACKOFF_DELAY_MS, multiplier = 2))
+  public Copy copy(CopyObjectRequest copyObjectRequest, TransferStateChangeListener stateChangeListener) {
+    retryAttempt++;
+    LOG.info("copying attempt {}/3", retryAttempt);
```
Comment (patduin) [inline circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java]: Where does this 3 come from, is that a spring default?

File: circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import com.amazonaws.services.s3.transfer.TransferManager;
+import com.amazonaws.services.s3.transfer.internal.TransferStateChangeListener;
+
+public class RetryableTransferManager {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RetryableTransferManager.class);
+  private static final int BACKOFF_DELAY_MS = 500;
+  private TransferManager transferManager;
+  private AmazonS3 srcClient;
+  private int retryAttempt = 0;
```
Comment (patduin) [inline circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java]: should we start counting at 1 more human readable to read attempt 1/3, 2/3, 3/3 instead of 0/3, 1/3, 2/3?

File: circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java (hunk 1 above)
Comment (max-jacobs) [reply circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java]: Yep Spring defaults to 3 attempts. I've also noticed that if there are multiple objects being copied, the `retryAttempt` will continue to increment. Adding a test for this now so that the logs are correct.

File: circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    this.transferManager = transferManager;
+    this.srcClient = srcClient;
+  }
+
+  @Retryable(value = { AmazonS3Exception.class },
+      backoff = @Backoff(delay = BACKOFF_DELAY_MS, multiplier = 2))
+  public Copy copy(CopyObjectRequest copyObjectRequest, TransferStateChangeListener stateChangeListener) {
+    retryAttempt++;
+    LOG.info("copying attempt {}/3", retryAttempt);
+    return transferManager.copy(copyObjectRequest, srcClient, stateChangeListener);
```
Comment (patduin) [inline circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java]: is it ok to reuse the client on every attempt or should we create a new one?

File: circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  private int retryAttempt = 0;
+
+  public RetryableTransferManager(TransferManager transferManager, AmazonS3 srcClient) {
+    this.transferManager = transferManager;
+    this.srcClient = srcClient;
+  }
+
+  @Retryable(value = { AmazonS3Exception.class },
+      backoff = @Backoff(delay = BACKOFF_DELAY_MS, multiplier = 2))
+  public Copy copy(CopyObjectRequest copyObjectRequest, TransferStateChangeListener stateChangeListener) {
```
Comment (patduin) [inline circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java]: Can we configure the number of attempts to try??

File: circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java (hunk 1 above)
Comment (patduin) [reply circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java]: ok, let's make the nr of attempts configurable. I guess we can't use the spring property as that would be the same for all RetryTemplates (we'd have a clash if we ever add another in the future??).

Comment (max-jacobs) [reply circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java]: Do you mean make it configurable for the user? If so I will have to refactor as annotation values must be constants and so aren't configurable.

And yes, with this setup, the spring property would apply across all retries. It may be possible to maintain multiple `RetryTemplate`s to avoid having clashes.

File: circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java (hunk 2 above)
Comment (max-jacobs) [reply circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java]: The logging count does indeed start at 1, the increment occurs just before the log each time :)

Comment (patduin) [reply circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java]: ah of course my bad

File: circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java (hunk 3 above)
Comment (max-jacobs) [reply circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java]: This is configured in the same way as the current setup. The client is instantiated at the very beginning when the copy jobs are started and doesn't change throughout this process. Any new calls to the top level `copy()` method in `S3S3Copier` will create a new client.

File: circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java (hunk 1 above)
Comment (abhimanyugupta07) [reply circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java]: Also, it would be nice to declare '3' as a constant (perhaps, MAX_RETRIES).

Comment (max-jacobs) [reply circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java]: @abhimanyugupta07 I've gone with `MAX_ATTEMPTS` as it includes the first failure :)

File: circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/S3S3Copier.java [hunk 5]
```diff
@@ -156,7 +156,7 @@ private void startAllCopyJobs() {
     AmazonS3URI targetBase = toAmazonS3URI(replicaLocation.toUri());
     srcClient = s3ClientFactory.newInstance(sourceBase, s3s3CopierOptions);
     targetClient = s3ClientFactory.newInstance(targetBase, s3s3CopierOptions);
-    transferManager = transferManagerFactory.newInstance(targetClient, s3s3CopierOptions);
+    retryableTransferManager = retryableTransferManagerFactory.newInstance(targetClient, s3s3CopierOptions, srcClient);
```
Comment (abhimanyugupta07) [inline circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/S3S3Copier.java]: I see that `newInstance` is taking in `srcClient` as the argument now. Do we know why was it not like this in `TransferManagerFactory`. Just want to understand, if there was any specific reason for not tying up the transferManager to a sourceClient. If this coupling is unnecessary, than I think, it should be avoided.

@patduin thoughts ?

File: circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManagerFactory.java [hunk 6]
```diff
@@ -24,15 +24,16 @@
 import com.hotels.bdp.circustrain.s3s3copier.S3S3CopierOptions;
 
 @Component
-public class TransferManagerFactory {
+public class RetryableTransferManagerFactory {
```
Comment (abhimanyugupta07) [inline circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManagerFactory.java]: Why is this change necessary ? A transferManagerFactory is a Factory which returns different types of transferManagers, in your case now, you will return a RetryableTransferManager.

Please let me know if I am missing anything.

File: circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java (hunk 1 above)
Comment (abhimanyugupta07) [reply circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java]: And use that constant on line 49 in the log statment as well, please.

Comment (max-jacobs) [reply circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java]: ah yes, good spot, thanks!

Comment (patduin) [reply circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java]: this doesn't work: ``` maxAttempts="${s3toS3copier-retry-maxattempts}"``` with that value coming from the yml? Might have to use: `maxAttemptsExpression`

Comment (max-jacobs) [reply circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java]: It doesn't look like it unfortunately, Java annotations must be constants. Have I understood you correctly? 

As a workaround, I could hardcode the default number of attempts and then break from the retry early if the user specifies a lower number of attempts?

File: circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java (hunk 3 above)
Comment (patduin) [reply circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java]: ok cool, should be fine then!

File: circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/S3S3Copier.java (hunk 5 above)
Comment (patduin) [reply circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/S3S3Copier.java]: it was already tied just went from `copy` see line 214 to the`newInstance` method. That's ok it is a field so makes more sense to load it in here. The copy method now has all arguments with the same scope.

File: circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java (hunk 4 above)
Comment (max-jacobs) [reply circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java]: Yep I have updated this now

File: circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManagerFactory.java (hunk 6 above)
Comment (max-jacobs) [reply circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManagerFactory.java]: Hey, can I resolve this conversation? Thanks!

Comment (abhimanyugupta07) [reply circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManagerFactory.java]: Yes Please.

File: circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java (hunk 1 above)
Comment (max-jacobs) [reply circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java]: This has now been refactored so that the `maxAttempts` property is configurable

File: circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/S3S3CopierOptions.java [hunk 7]
```diff
@@ -54,7 +54,11 @@
      * {@link com.amazonaws.services.s3.model.CannedAccessControlList}
      * Optional AWS Canned ACL
      */
-    CANNED_ACL("canned-acl");
+    CANNED_ACL("canned-acl"),
+    /**
+     * Number of copy attempts to allow when copying from S3 to S3. Default value is 1.
```
Comment (patduin) [inline circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/S3S3CopierOptions.java]: Shall we just do 3 as a default? No harm right usually no retry is necesary

File: circus-train-s3-s3-copier/src/test/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManagerTest.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+import com.amazonaws.services.s3.transfer.internal.TransferStateChangeListener;
+
+@RunWith(SpringJUnit4ClassRunner.class)
+@ContextConfiguration(classes = RetryableTransferManagerTest.ContextConfiguration.class)
+public class RetryableTransferManagerTest {
+
+  private static int MAX_COPY_ATTEMPTS = 3;
+
+  @Autowired
+  private TransferManager mockedTransferManager;
```
Comment (patduin) [inline circus-train-s3-s3-copier/src/test/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManagerTest.java]: What do we gain but injecting the classes via spring instead of mocking them directly?

File: circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/S3S3CopierOptions.java (hunk 7 above)
Comment (max-jacobs) [reply circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/S3S3CopierOptions.java]: Sure thing, that makes sense

File: circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+  public Copy copy(final CopyObjectRequest copyObjectRequest, final TransferStateChangeListener stateChangeListener) {
+    try {
+      return retryTemplate.execute(new RetryCallback<Copy, Throwable>() {
+        public Copy doWithRetry(RetryContext context) {
+          LOG.info("copying attempt {}/{}", context.getRetryCount()+1, maxAttempts);
+          return transferManager.copy(copyObjectRequest, srcClient, stateChangeListener);
+        }
+      });
+    } catch (Throwable throwable) {
+      throw new AmazonS3Exception(throwable.getMessage());
```
Comment (patduin) [inline circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java]: We should add the throwable as the cause otherwise we loose the original stacktrace

File: circus-train-s3-s3-copier/src/test/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManagerTest.java (hunk 8 above)
Comment (max-jacobs) [reply circus-train-s3-s3-copier/src/test/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManagerTest.java]: Yep these can be mocked directly, and the mockito runner can be used as well. I was playing around with autowiring spring properties but should have changed it back. Thanks!

File: circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java (hunk 9 above)
Comment (max-jacobs) [reply circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java]: Good idea!

File: circus-train-s3-s3-copier/src/test/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManagerTest.java (hunk 8 above)
Comment (patduin) [reply circus-train-s3-s3-copier/src/test/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManagerTest.java]: 👍 always good to try out stuff!

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,3 +1,8 @@
+## 13.2.2 - 2019-02-20
+### Changed
+* Added a retry mechanism to handle AWS S3 copying bug
+* Added a configuration parameter to support a configurable number of maximum attempts at copying
```
Comment (patduin) [inline CHANGELOG.md]: this line can go, enough we just mention the retry I think. Can you add a link to the issue see examples below

File: CHANGELOG.md [hunk 11]
```diff
@@ -1,3 +1,8 @@
+## 13.2.2 - 2019-02-20
+### Changed
```
Comment (abhimanyugupta07) [inline CHANGELOG.md]: I think this should be `Added` rather than `Changed`
and then you can remove the word 'Added' from the start of statement.

https://keepachangelog.com/en/1.0.0/

File: CHANGELOG.md (hunk 10 above)
Comment (abhimanyugupta07) [reply CHANGELOG.md]: Or make it 
```Added a configurable retry mechanism to handle flaky AWS S3 to S3 copying```.

Comment (max-jacobs) [reply CHANGELOG.md]: Thanks both, will make this more concise

File: circus-train-s3-s3-copier/src/test/java/com/hotels/bdp/circustrain/s3s3copier/S3S3CopierOptionsTest.java [hunk 12]
```diff
@@ -96,4 +96,24 @@ public void getCannedAclDefaultIsNull() throws Exception {
...
 
+  @Test
+  public void getMaxCopyAttempts() throws Exception {
+    copierOptions.put(S3S3CopierOptions.Keys.MAX_COPY_ATTEMPTS.keyName(), 3);
+    S3S3CopierOptions options = new S3S3CopierOptions(copierOptions);
+    assertThat(options.getMaxCopyAttempts(), is(3));
+  }
+
+  @Test
+  public void getMaxCopyAttemptsDefaultIsOne() throws Exception {
```
Comment (abhimanyugupta07) [inline circus-train-s3-s3-copier/src/test/java/com/hotels/bdp/circustrain/s3s3copier/S3S3CopierOptionsTest.java]: When we change the defaults to Three, please rename this test.

File: CHANGELOG.md (hunk 11 above)
Comment (max-jacobs) [reply CHANGELOG.md]: Thanks for the link, looks interesting. Will update now.

File: circus-train-s3-s3-copier/src/test/java/com/hotels/bdp/circustrain/s3s3copier/S3S3CopierOptionsTest.java [hunk 13]
```diff
@@ -96,4 +96,24 @@ public void getCannedAclDefaultIsNull() throws Exception {
...
+    S3S3CopierOptions options = new S3S3CopierOptions(copierOptions);
+    assertThat(options.getMaxCopyAttempts(), is(3));
+  }
+
+  @Test
+  public void getMaxCopyAttemptsDefaultIsOneIfLessThanOne() throws Exception {
```
Comment (abhimanyugupta07) [inline circus-train-s3-s3-copier/src/test/java/com/hotels/bdp/circustrain/s3s3copier/S3S3CopierOptionsTest.java]: This one needs renaming too.

Comment (abhimanyugupta07) [reply circus-train-s3-s3-copier/src/test/java/com/hotels/bdp/circustrain/s3s3copier/S3S3CopierOptionsTest.java]: Shouldn't there be a validation check at the config level that the numAttempts can't be less than one ?

File: circus-train-integration-tests/src/test/data/com/hotels/bdp/circustrain/integration/CircusTrainS3S3IntegrationTest/partitioned-single-table-with-no-partitions-mirror.yml [hunk 14]
```diff
@@ -8,3 +8,5 @@ table-replications:
       table-name: ct_table_p_copy
 security:
   credential-provider: jceks://file/${config-location}/aws.jceks
+copier-options:
+  s3s3copier-retry-maxattempts: 3
```
Comment (massdosage) [inline circus-train-integration-tests/src/test/data/com/hotels/bdp/circustrain/integration/CircusTrainS3S3IntegrationTest/partitioned-single-table-with-no-partitions-mirror.yml]: We might want to consider renaming this to s3-copy-retry-maxattempts so we can use the same option for both the S3S3Copier and the S3MapReduceCopier (if we decide to add a similar retry feature to the latter).

Comment (massdosage) [reply circus-train-integration-tests/src/test/data/com/hotels/bdp/circustrain/integration/CircusTrainS3S3IntegrationTest/partitioned-single-table-with-no-partitions-mirror.yml]: Or if we don't add it to the S3MapReduceCopier we should rename this to `s3s3-retry-max-attempts` to be more consistent with the other s3s3 specific options (e.g. `s3s3-multipart-copy-threshold-in-bytes`, `s3s3-multipart-copy-part-size-in-bytes`).

File: circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+import com.amazonaws.services.s3.model.AmazonS3Exception;
+import com.amazonaws.services.s3.model.CopyObjectRequest;
+import com.amazonaws.services.s3.transfer.Copy;
+import com.amazonaws.services.s3.transfer.TransferManager;
+import com.amazonaws.services.s3.transfer.internal.TransferStateChangeListener;
+
+public class RetryableTransferManager {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RetryableTransferManager.class);
+  private static final int INITIAL_RETRY_INTERVAL_MS = 500;
```
Comment (massdosage) [inline circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java]: Let's go for slightly higher values here, it's not entirely clear what the error is on the AWS side but I have a feeling we should wait a bit longer for it to resolve itself. So let's go for 2000 for initial and 6000 for max.

File: circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+    this.maxAttempts = maxAttempts;
+    this.retryTemplate = setUpRetryTemplate();
+  }
+
+  public Copy copy(final CopyObjectRequest copyObjectRequest, final TransferStateChangeListener stateChangeListener) {
+    try {
+      return retryTemplate.execute(new RetryCallback<Copy, Throwable>() {
+        public Copy doWithRetry(RetryContext context) {
+          LOG.info("copying attempt {}/{}", context.getRetryCount()+1, maxAttempts);
+          return transferManager.copy(copyObjectRequest, srcClient, stateChangeListener);
```
Comment (massdosage) [inline circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java]: As discussed, the error might not actually be throw here it might be later on when we go through the list of submitted copy jobs and wait for them to complete.

File: circus-train-integration-tests/src/test/data/com/hotels/bdp/circustrain/integration/CircusTrainS3S3IntegrationTest/partitioned-single-table-with-no-partitions-mirror.yml (hunk 14 above)
Comment (max-jacobs) [reply circus-train-integration-tests/src/test/data/com/hotels/bdp/circustrain/integration/CircusTrainS3S3IntegrationTest/partitioned-single-table-with-no-partitions-mirror.yml]: Seeing as the map reduce copier has a retry already built in, shall we go with `s3s3-retry-max-copy-attempts` for consistency with the other options for now?

File: circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java (hunk 16 above)
Comment (max-jacobs) [reply circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java]: Hmm, in this case I can add some logging so we can capture this more clearly.

The better solution would be to refactor the retry logic so that we can retry the job if it fails while it is completing - either by introducing some parallelism, or by caching the jobs as they are submitted and retrying if they are discovered to have failed when we inspect their progress.

File: circus-train-integration-tests/src/test/data/com/hotels/bdp/circustrain/integration/CircusTrainS3S3IntegrationTest/partitioned-single-table-with-no-partitions-mirror.yml (hunk 14 above)
Comment (massdosage) [reply circus-train-integration-tests/src/test/data/com/hotels/bdp/circustrain/integration/CircusTrainS3S3IntegrationTest/partitioned-single-table-with-no-partitions-mirror.yml]: Yes.

File: circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java (hunk 16 above)
Comment (massdosage) [reply circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/RetryableTransferManager.java]: Agreed, let's see if AWS respond to the case we have open with them for the underlying issue on their side and we can then decide on next steps.
